core/src/main/java/io/questdb/cutlass/pgwire/PGConnectionContext.java

```diff
@@ -1344,19 +1344,27 @@ private void executeTag0() {
         }
     }
 
-    private void executeUpdate() throws SqlException {
-        try {
-            if (transactionState != ERROR_TRANSACTION) {
-                // when transaction is in error state, skip execution
-                executeUpdate0();
-            }
-            prepareCommandComplete(true);
-        } catch (Throwable e) {
-            if (transactionState == IN_TRANSACTION) {
-                transactionState = ERROR_TRANSACTION;
+    private void executeUpdate(SqlCompiler compiler) throws SqlException {
+        boolean recompileStale = true;
+        do {
+            try {
+                if (transactionState != ERROR_TRANSACTION) {
+                    // when transaction is in error state, skip execution
+                    executeUpdate0();
+                    recompileStale = false;
+                }
+                prepareCommandComplete(true);
+            } catch (ReaderOutOfDateException e) {
+                LOG.info().$(e.getFlyweightMessage()).$();
+                typesAndUpdate = Misc.free(typesAndUpdate);
+                compileQuery(compiler);
```

Comment:
is this a recursion? There are similar places that encounter this exception but they all try to deal with infinite recursion.

Reply:
I don't think there is a recursion. the `ReaderOutOfDateException` is thrown during Operation execution, not during compilation. `compileQuery()` is not calling this method. Unless I'm missing something.

Reply:
what can happen is that `compileQuery()` will produce a plan which will be outdated at the moment it's executed. is that what you meant?

Reply:
Not a recursion, but a potentially infinite loop. Though the probability of infinite spins is close to 0. Potentially, we could restrict the total number of retries if you think it's safer this way.

Reply:
I agree, this is a potential infinite loop. In fact multiple sites do the similar things and don't deal with infinite loops. @jerrinot could you review other sites of handing of this exception please.

On a separate note:

```
compileQuery(compiler) -> processCompiledQuery(compiler.compile(queryText, sqlExecutionContext));
```

Reply:
ok, I guess we could/should limit the number of recompilation attempts. I will have a look at this.

Reply:
I see this is solved now